File: src/core/passes/shared-call-tail.ts
import type { IrOp } from "../types.ts";
import { hasRewriteBarrier, type IrPass, type IrPassFn } from "./helpers.ts";
import { createLabelAllocator, targetKey } from "./outline.ts";

interface SharedCallTail {
  key: string;
  label: string;
  call: Extract<IrOp, { kind: "call" }>;
  continuation: Extract<IrOp, { kind: "jump" }>;
  count: number;
}

const run: IrPassFn = (ops) => {
  const candidates = collectSharedCallTails(ops);
  const normalizeContinuation = buildContinuationNormalizer(ops);
  if (candidates.size === 0) return { ops: [...ops], applied: 0, optimizations: [] };

  const result: IrOp[] = [];
  let applied = 0;
  for (let index = 0; index < ops.length; index += 1) {
    const op = ops[index]!;
    const next = ops[index + 1];
    if (op.kind === "call" && next?.kind === "jump") {
      const key = callTailKey(op, next, normalizeContinuation);
      const candidate = candidates.get(key);
      if (candidate !== undefined) {
        result.push({
          kind: "jump",
          target: candidate.label,
          opcode: 0x51,
          meta: {
            ...op.meta,
            mnemonic: "БП",
            comment: op.meta.comment?.replace(/^proc call/u, "shared call tail") ?? "shared call tail",
          },
          targetMeta: { comment: "shared call tail" },
        });
        index += 1;
        applied += 1;
        continue;
      }
    }
    result.push(op);
  }

  for (const candidate of candidates.values()) {
    result.push({ kind: "label", name: candidate.label });
    result.push({
      ...candidate.call,
      meta: {
        ...candidate.call.meta,
        comment: candidate.call.meta.comment ?? "shared call tail helper",
      },
    });
    result.push({
      ...candidate.continuation,
      meta: {
        ...candidate.continuation.meta,
        comment: candidate.continuation.meta.comment ?? "shared call tail continuation",
      },
    });
  }

  return {
    ops: result,
    applied,
    optimizations: [{
      name: "shared-call-tail",
      detail: `Shared ${applied} call+jump tail sequence${applied === 1 ? "" : "s"}.`,
    }],
  };
};

export const sharedCallTail: IrPass = {
  name: "shared-call-tail",
  run,
  layoutSafe: false,
};

function collectSharedCallTails(ops: readonly IrOp[]): Map<string, SharedCallTail> {
  const normalizeContinuation = buildContinuationNormalizer(ops);
  const counts = new Map<string, {
    call: Extract<IrOp, { kind: "call" }>;
    continuation: Extract<IrOp, { kind: "jump" }>;
    count: number;
  }>();
  for (let index = 0; index < ops.length - 1; index += 1) {
    const op = ops[index]!;
    const next = ops[index + 1]!;
    if (op.kind !== "call" || next.kind !== "jump") continue;
    if (hasRewriteBarrier(op) || hasRewriteBarrier(next)) continue;
    const normalizedTarget = normalizeContinuation(next.target);
    const continuation: Extract<IrOp, { kind: "jump" }> = {
      ...next,
      target: normalizedTarget,
    };
    const key = callTailKey(op, continuation, normalizeContinuation);
    const existing = counts.get(key);
    if (existing === undefined) {
      counts.set(key, { call: op, continuation, count: 1 });
    } else {
      existing.count += 1;
    }
  }

  const result = new Map<string, SharedCallTail>();
  const labels = createLabelAllocator(ops, "__shared_call_tail_");
  for (const [key, candidate] of counts) {
    if (candidate.count < 3) continue;
    result.set(key, {
      key,
      label: labels.next(),
      call: candidate.call,
      continuation: candidate.continuation,
      count: candidate.count,
    });
  }
  return result;
}

function callTailKey(
  call: Extract<IrOp, { kind: "call" }>,
  continuation: Extract<IrOp, { kind: "jump" }>,
  normalizeContinuation: (target: string | number) => string | number,
): string {
  return `${targetKey(call.target)}|${targetKey(normalizeContinuation(continuation.target))}`;
}

function buildContinuationNormalizer(ops: readonly IrOp[]): (target: string | number) => string | number {
  const labelIndexes = new Map<string, number>();
  for (let index = 0; index < ops.length; index += 1) {
    const op = ops[index]!;
    if (op.kind === "label") labelIndexes.set(op.name, index);
  }

  const normalize = (target: string | number, seen = new Set<string>()): string | number => {
    if (typeof target !== "string") return target;
    if (seen.has(target)) return target;
    seen.add(target);
    const labelIndex = labelIndexes.get(target);
    if (labelIndex === undefined) return target;
    const executableIndex = nextExecutableIndex(ops, labelIndex + 1);
    const executable = executableIndex === undefined ? undefined : ops[executableIndex];
    if (executable?.kind !== "jump" || hasRewriteBarrier(executable)) return target;
    return normalize(executable.target, seen);
  };

  return (target) => normalize(target);
}

function nextExecutableIndex(ops: readonly IrOp[], start: number): number | undefined {
  for (let index = start; index < ops.length; index += 1) {
    if (ops[index]?.kind !== "label") return index;
  }
  return undefined;
}

File: native/src/core/passes/shared_call_tail.cpp
#include "mkpro/core/passes/shared_call_tail.hpp"

#include "mkpro/core/passes/outline.hpp"

#include <map>
#include <optional>
#include <set>
#include <string>
#include <utility>
#include <variant>
#include <vector>

namespace mkpro::core::passes {

namespace {

struct SharedCallTail {
  std::string key;
  std::string label;
  IrOp call;
  IrOp continuation;
  int count = 0;
};

std::map<std::string, int> build_label_indexes(const std::vector<IrOp>& ops) {
  std::map<std::string, int> label_indexes;
  for (int index = 0; index < static_cast<int>(ops.size()); ++index) {
    const IrOp& op = ops.at(static_cast<std::size_t>(index));
    if (op.kind == IrKind::Label)
      label_indexes[op.name] = index;
  }
  return label_indexes;
}

IrTarget normalize_continuation(const std::vector<IrOp>& ops,
                                const std::map<std::string, int>& label_indexes,
                                const IrTarget& target, std::set<std::string> seen = {}) {
  const auto* label = std::get_if<std::string>(&target);
  if (label == nullptr)
    return target;
  if (seen.contains(*label))
    return target;
  seen.insert(*label);

  const auto label_index = label_indexes.find(*label);
  if (label_index == label_indexes.end())
    return target;
  const std::optional<int> executable_index = next_executable_index(ops, label_index->second + 1);
  if (!executable_index.has_value())
    return target;
  const IrOp& executable = ops.at(static_cast<std::size_t>(*executable_index));
  if (executable.kind != IrKind::Jump || has_rewrite_barrier(executable))
    return target;
  return normalize_continuation(ops, label_indexes, executable.target, std::move(seen));
}

std::string call_tail_key(const IrOp& call, const IrOp& continuation,
                          const std::vector<IrOp>& ops,
                          const std::map<std::string, int>& label_indexes) {
  return target_key(call.target) + "|" +
         target_key(normalize_continuation(ops, label_indexes, continuation.target));
}

std::map<std::string, SharedCallTail> collect_shared_call_tails(const std::vector<IrOp>& ops) {
  const std::map<std::string, int> label_indexes = build_label_indexes(ops);
  struct CountedTail {
    IrOp call;
    IrOp continuation;
    int count = 0;
  };
  std::map<std::string, CountedTail> counts;

  for (int index = 0; index < static_cast<int>(ops.size()) - 1; ++index) {
    const IrOp& op = ops.at(static_cast<std::size_t>(index));
    const IrOp& next = ops.at(static_cast<std::size_t>(index + 1));
    if (op.kind != IrKind::Call || next.kind != IrKind::Jump)
      continue;
    if (has_rewrite_barrier(op) || has_rewrite_barrier(next))
      continue;

    IrOp continuation = next;
    continuation.target = normalize_continuation(ops, label_indexes, next.target);
    const std::string key = call_tail_key(op, continuation, ops, label_indexes);
    auto found = counts.find(key);
    if (found == counts.end()) {
      counts[key] = CountedTail{.call = op, .continuation = continuation, .count = 1};
    } else {
      found->second.count += 1;
    }
  }

  std::map<std::string, SharedCallTail> result;
  LabelAllocator labels(ops, "__shared_call_tail_");
  for (const auto& [key, candidate] : counts) {
    if (candidate.count < 3)
      continue;
    result[key] = SharedCallTail{
        .key = key,
        .label = labels.next(),
        .call = candidate.call,
        .continuation = candidate.continuation,
        .count = candidate.count,
    };
  }
  return result;
}

std::optional<std::string> shared_call_comment(const std::optional<std::string>& comment) {
  if (!comment.has_value())
    return "shared call tail";
  if (comment->starts_with("proc call")) {
    return std::string("shared call tail") + comment->substr(std::string("proc call").size());
  }
  return *comment;
}

} // namespace

PassResult shared_call_tail(const std::vector<IrOp>& ops, const PassContext& context) {
  (void)context;
  const std::map<std::string, SharedCallTail> candidates = collect_shared_call_tails(ops);
  const std::map<std::string, int> label_indexes = build_label_indexes(ops);
  if (candidates.empty())
    return PassResult{.ops = ops, .applied = 0, .optimizations = {}};

  std::vector<IrOp> result;
  result.reserve(ops.size());
  int applied = 0;

  for (int index = 0; index < static_cast<int>(ops.size()); ++index) {
    const IrOp& op = ops.at(static_cast<std::size_t>(index));
    const IrOp* next =
        index + 1 < static_cast<int>(ops.size()) ? &ops.at(static_cast<std::size_t>(index + 1))
                                                 : nullptr;
    if (op.kind == IrKind::Call && next != nullptr && next->kind == IrKind::Jump) {
      const std::string key = call_tail_key(op, *next, ops, label_indexes);
      const auto candidate = candidates.find(key);
      if (candidate != candidates.end()) {
        IrOp replacement;
        replacement.kind = IrKind::Jump;
        replacement.target = candidate->second.label;
        replacement.opcode = 0x51;
        replacement.meta = op.meta;
        replacement.meta.mnemonic = "БП";
        replacement.meta.comment = shared_call_comment(op.meta.comment);
        replacement.target_meta.comment = "shared call tail";
        result.push_back(std::move(replacement));
        ++index;
        ++applied;
        continue;
      }
    }
    result.push_back(op);
  }

  for (const auto& [unused_key, candidate] : candidates) {
    (void)unused_key;
    IrOp label;
    label.kind = IrKind::Label;
    label.name = candidate.label;
    result.push_back(std::move(label));

    IrOp call = candidate.call;
    if (!call.meta.comment.has_value())
      call.meta.comment = "shared call tail helper";
    result.push_back(std::move(call));

    IrOp continuation = candidate.continuation;
    if (!continuation.meta.comment.has_value())
      continuation.meta.comment = "shared call tail continuation";
    result.push_back(std::move(continuation));
  }

  return PassResult{
      .ops = std::move(result),
      .applied = applied,
      .optimizations =
          {
              AppliedOptimization{
                  .name = "shared-call-tail",
                  .detail = "Shared " + std::to_string(applied) + " call+jump tail sequence" +
                            (applied == 1 ? "" : "s") + ".",
              },
          },
  };
}

IrPass shared_call_tail_pass() {
  return IrPass{
      .name = "shared-call-tail",
      .run = shared_call_tail,
      .layout_safe = false,
  };
}

} // namespace mkpro::core::passes
